### ISSUE TITLE
Support scoping inner elements of a webc file 

### ISSUE THREAD

**Issue by bglw** [11ty/webc#5]
## Assumption about the current state:

If I have the file:
```html
<div>
  <h1>Example</h1>
  <p>Content</p>
</div>

<style webc:scoped>
:host h1 {
	color: aliceblue;
}
:host p {
	color: blanchedalmond;
}
</style>
```

After compilation I would currently see:
```js
{
	html: "<div class=\"wcl2xedjk\"><h1>Example</h1><p>Content</p></div>",
	css: [".wcl2xedjk h1{color:aliceblue}.wcl2xedjk p{color:blanchedalmond}"],
}
```

## Main concern:

For components that wrap another component, this isn't scoped enough to prevent style leakage. For example if our component above had a child component within it, and that child component also contained an `h1` element, then `.wcl2xedjk h1` would apply and turn it `aliceblue`.

## Desired behaviour:

In Svelte's CSS scoping, I can write the following:

```html
<div>
  <h1>Example</h1>
  <p>Content</p>
</div>

<style>
h1 {
	color: aliceblue;
}
p {
	color: blanchedalmond;
}
</style>
```

Which would provide me something along the psuedo-lines of
```js
{
	html: "<div><h1 class="kziuyb">Example</h1><p class="wgrhia">Content</p></div>",
	css: ["h1.kziuyb{color:aliceblue}p.wgrhia{color:blanchedalmond}"],
}
```

Where the styles are scoped directly to the elements they applied to.

## Implementation:

In (my) ideal world this would happen automatically when `webc:scoped` is applied. To keep it opt-in it one could utilize something like a `:host` pseudo-class wrapping styles:

```html
<style webc:scoped>
:host(h1) {
	color: aliceblue;
}
:host(p) {
	color: blanchedalmond;
}
</style>
```

Ideally slightly more complex selectors would also be handled, so `h1 + p:first-child` could become `h1.qwerty + p.asdfgh:first-child`

**Comment by zachleat** [comment]
Reclassifying this one as a bug

**Comment by zachleat** [comment]
I’m waffling back and forth here because I do kinda think this is _okay_ and as a better short term solution I’d like to ship an example of how to override the `webc:scoped` behavior with any arbitrary css scoping library e.g. css modules maybe? Any opinions on what a good next-level library for this might be?

**Comment by bglw** [comment]
Makes sense. I don't know of any css scoping libraries myself, but that would be a cool thing to find

**Comment by mayank99** [comment]
[`lightningcss`](https://lightningcss.dev/) might be a good library to show how to use css modules with. If not, [`postcss-modules`](https://github.com/madyankin/postcss-modules) is another option (i believe this is what vue uses under-the-hood as well). both of these solve the inner-element scoping problem and they give you a json mapping of the original vs transformed class name, which could be substituted against own webc's internal mapping.

also worth mentioning that [native scoping](https://drafts.csswg.org/css-nesting/) is coming and it might be good to start thinking about it. for example, webc could manage a unique hash to pass inside `@scope`.

**Comment by riskygit** [comment]
This is still important as long as scoped prefixing occurs before transforming... currently, the CSS being fed to a preprocessor is not what the developer wrote, but the result of applying the prefix transform. This creates unintended side effects.

**Comment by bolu61** [comment]
Is there currently a workaround to prevent styles from leaking into subcomponents?
